### ISSUE TITLE
Bad interaction of --depth and --flat: accounts missing (BZ#1020)

### ISSUE THREAD

**Issue by tbm** [ledger/ledger#1020]
Note: the issue was created automatically with bugzilla2github

Bugzilla bug ID: BZ#1020
From: Martin Michlmayr (@tbm)
CC:   @simonmichael, @thdox

**Comment by scarpent** [comment]
Seeing this as well: https://groups.google.com/forum/?fromgroups=#!topic/ledger-cli/pHhIXIT5KPo

**Comment by jwiegley** [comment]
I think `--flat` is ignoring all logic about summing accounts into their parent based on the cutoff depth, and instead is just treating the collection of accounts as exactly what they are. Clearly this logic is not what is expected.

**Comment by narendraj9** [comment]
I am planning to switch to `ledger` from `hledger` but this is something that I have gotten used to. I will try to see if I can find the fix for it.

**Comment by TwoD** [comment]
Is `--depth` working as it should when used on `register`? Seems all transactions are excluded the same way there too.
I would have expected to be able to have a large hierarchy of accounts and then be able to get a broader view of where they're happening using `--depth`. With the test case above `ledger --depth 2 reg` completely drops `Assets:BBB:CCC` from the transaction list instead of showing it as `Assets:BBB` as I would expect. Is my assumption wrong and/or is this a separate issue from this one?

**Comment by tbm** [comment]
I thought PR #1901 would fix this but apparently not. @Rahix since you know this code now, do you think you can take a look?

**Comment by Rahix** [comment]
PR #1901 was only for register reports so it makes sense the balance report is still broken.  I.e. this issue is separate from what I was working on.  But if I can find some time in the next weeks, I'll take a look!

**Comment by Rahix** [comment]
[`opt-depth_flat.test`](https://github.com/ledger/ledger/blob/master/test/baseline/opt-depth_flat.test) seems to test exactly the weird behavior described here.  I can't find history why it was added?  But the consensus here is that it should be changed (and thus the test with it as well), right?

**Comment by scarpent** [comment]
👍 from me, for what it's worth. 😀 

Thank you for working on this!

**Comment by porridge** [comment]
I hit this today trying to export amounts and account names from a balance report with the goal of making a pie chart.

In case others find it useful, it looks like `--balance-format` can be used as a workaround.

For my use case, `--depth 2 --balance-format='%(account),%(quantity(display_total))\n' --no-total balance` did the trick!

Thanks for making `ledger` so flexible!

**Comment by tbm** [comment]
@simonmichael 

> hledger does this. http://hledger.org/manual.html#flat-mode describes how flat mode and depth limiting interact.

I can't find this in the docs anymore. Do you have a new link/reference?

**Comment by simonmichael** [comment]
Yup, https://hledger.org/hledger.html#list-or-tree-mode and https://hledger.org/hledger.html#depth-limiting now.

**Comment by Rahix** [comment]
Just FYI: I had a stab at fixing this but it wasn't as easy as I had hoped - the problem is that sometimes new accounts (at the cut-off depth) need to be synthesized for this and so far I did not get it to work...  If someone else knows the codebase better, I'd love some hints where to dig - I don't have much time to immerse myself into it much deeper at the moment.

**Comment by tbm** [comment]
@Rahix can you email John directly and ask if he can do a pair programming session / walk through the problem on a video call?

**Comment by tbm** [comment]
Removing milestone since it's unlikely this will be fixed soon.
